src/index.js

```diff
@@ -49,32 +50,35 @@ function createFormAction (requestAction, types, payloadCreator = identity) {
   }, actionMethods);
 };
 
-function *formActionSaga () {
-  while (true) {
-    let action = yield take(PROMISE);
-    let { request, defer, types } = action.payload;
-    let { resolve, reject } = defer;
-    let [ SUCCESS, FAIL ] = types;
+function *handlePromiseSaga({ payload }) {
+  const { request, defer, types } = payload;
+  const { resolve, reject } = defer;
+  const [ SUCCESS, FAIL ] = types;
 
-    yield put(request);
-
-    const winner = yield race({
+  const [ winner ] = yield [
+    race({
       success: take(SUCCESS),
-      fail: take(FAIL)
-    });
+      fail: take(FAIL),
+    }),
+    put(request),
+  ];
 
-    if (winner.success) {
-      yield call(resolve, winner.success);
-    } else {
-      yield call(reject, winner.fail);
-    }
+  if (winner.success) {
+    yield call(resolve, winner.success);
+  } else {
+    yield call(reject, winner.fail);
   }
 }
 
+function *formActionSaga() {
+  yield call(takeEvery, PROMISE, handlePromiseSaga);
```

Comment:
For testing purposes, see https://github.com/yelouafi/redux-saga/issues/318#issuecomment-218092686